### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by ADmad**
The public API remains backward compatible so I think just minor version bump should suffice when making a new release.

### CODE REVIEW

**Comment by ADmad** [comment]
Should we add a new method `Manager::getFillterCollection()` which return `FilterCollectionInterface` and deprecate the `Manager::getFilters()` method?

**Comment by ADmad** [comment]
phpstan is unhappy since `getConfig()` is called on filter instance but `FilterInterface` doesn't have that method :unamused:

**Comment by dereuromark** [comment]
annotate for now till next major?

**Comment by ADmad** [comment]
Can't probably just add the method to `FilterInterface`. All existing filter would be extending `Base` filter which uses `InstanceConfigTrait` and hence would have `getConfig()` implemented.

**Comment by ADmad** [comment]
This shows having an interface as well as base class doesn't add much value, just adds pain :)

**Comment by dereuromark** [comment]
In this specific case, having an abstract class to work with and typehint against, an interface contracting approach is indeed not adding too much value.

**Comment by ADmad** [comment]
I'll remove the interface. No one is going to directly implement the interface instead of extending the `Base` filter.

**Comment by ADmad** [comment]
To facilitate lazy coding I have updated `SearchBehavior` to conventionally use a collection with name matching the table alias if it exists. So for example if you attach search behavior to `PostsTable` then by default `App\Model\Filter\PostsCollection` will be used if it exists.

File: README.md [hunk 1]
```diff
@@ -106,68 +78,135 @@ class ExampleTable extends Table
...
+
+Apart from configuring filters through search mananger in your table class,
+you can also create them as separate collection classes. This helps in
+keeping your table's `initialize()` method uncluttered and the filters are lazy
+loaded only when actually used.
 
-        return $searchManager;
+```php
+// src/Model/Filter/Filter/PostsCollection.php
+namespace App\Model\Filter\PostsCollection;
```
Comment (dereuromark) [inline README.md]: Is the namespace not just ...\Filter ?

File: README.md [hunk 2]
```diff
@@ -228,12 +248,12 @@ After including the trait you can use the searchManager by calling the
 
 ## Filtering your data
 Once you have completed all the setup you can now filter your data by passing
-query params in your index method. Using the `Article` example given above, you
-could filter your articles using the following.
+query params in your index method. Using theexample given above, you
```
Comment (dereuromark) [inline README.md]: space missing

File: src/Manager.php [hunk 3]
```diff
@@ -61,203 +78,110 @@ public function getRepository()
...
-     * @return $this
-     */
-    public function like($name, array $config = [])
-    {
-        $this->add($name, 'Search.Like', $config);
+        $instance = new $className($this);
+        if (!$instance instanceof FilterCollectionInterface) {
+            throw new InvalidArgumentException(sprintf(
+                'The collection must be instance of FilterCollectionInterface. ' .
+                'Got instanceof "%s" instead',
```
Comment (dereuromark) [inline src/Manager.php]:
```suggestion
                'Got instance of "%s" instead',
```

File: src/Model/Filter/FilterCollection.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * FilterCollection
+ */
+class FilterCollection implements FilterCollectionInterface
+{
+    use FilterMethodsTrait;
+
+    /**
+     * @var array List of filter objects
```
Comment (dereuromark) [inline src/Model/Filter/FilterCollection.php]:
```suggestion
     * @var \Search\Model\Filter\Base[] List of filter objects
```

File: src/Model/Filter/Like.php [hunk 5]
```diff
@@ -147,8 +148,10 @@ protected function _formatWildcards($value)
     protected function _setEscaper()
     {
         if ($this->getConfig('escaper') === null) {
-            /** @var \Cake\Database\Query $query */
```
Comment (dereuromark) [inline src/Model/Filter/Like.php]: Is the below code still autocompletable in IDE without this? $query->getConnection()->getDriver() etc?

File: src/Model/SearchTrait.php [hunk 6]
```diff
@@ -129,7 +141,7 @@ protected function _extractParams($params, $filters)
      * ```
      *
      * @param array $params The parameters array to flatten.
-     * @param array $filters The array of filters with configuration
+     * @param \Search\Model\Filter\Base[] $filters The array of filters with configuration
      * @return array The flattened parameters array.
      */
     protected function _flattenParams($params, $filters)
```
Comment (dereuromark) [inline src/Model/SearchTrait.php]: Ideally we also typehint the array on such a protected methods to make sure they are of that type going in.

File: src/Manager.php [hunk 7]
```diff
@@ -32,14 +37,26 @@ class Manager
      */
     protected $_collection = 'default';
```
Comment (dereuromark) [inline src/Manager.php]: There is a lot of magic string usage around `'default'`. Could we make this a class const to more easily reuse it in other places? Especially when we compare against it, typos here can easily lead to hard to find dev errors.

**Comment by dereuromark** [comment]
Looks good to go.
Whats the release timeline/plan?

File: README.md (hunk 1 above)
Comment (dereuromark) [reply README.md]: Thats my point: PostsCollection in the namespace then?
I expected `App\Model\Filter` only.

Comment (dereuromark) [reply README.md]: Also the file name above does not match the namespace.

**Comment by ADmad** [comment]
> Whats the release timeline/plan?

This weekend :slightly_smiling_face: 

I am going to bump major version due to the significant amount of changes but most user should be update to new version without requiring any changes in their app code.

**Comment by dereuromark** [comment]
If this is bc i prefer a minor.
All plugins using this as dependency otherwise need adjusting again.

**Comment by ADmad** [comment]
From usage POV it is BC but the signature of `Manager::__construct()` has changed. Not sure if that's allowed in a minor by semver.

**Comment by dereuromark** [comment]
I think most of this is really internal. So a minor (compared to the fallout of majoring) is acceptable IMO.
A major triggers really a lot of plugins to be incompatible and disallowing for it until adjusted via patch manually.

**Comment by ADmad** [comment]
I want to make some technically non BC changes like making `Manager::getFilters()` return filter collection instance instead of array which would warrant a major version bump.

I don't think a new major would be a problem. Any app/plugin wanting to use the new filters collection feature would just have to update the plugin's version to `^5.0`, that's it.

**Comment by dereuromark** [comment]
Fair enough.
Are there any other BC breaking things we want to do or clean up then?

**Comment by dereuromark** [comment]
I was thinking about making some methods protected that should never have been public in the first place.
Please see

    Manager::loadFilter()

etc.

**Comment by ADmad** [comment]
> Are there any other BC breaking things we want to do or clean up then?

Can't remember off top of my head but I'll go through the code and see if there's anything else which needs cleanup.

> Manager::loadFilter()

That method is already removed on this branch :)

**Comment by dereuromark** [comment]
Yeah, I saw and updated my comment. Still think that method should not be exposed as public interface method, though :)

**Comment by ADmad** [comment]
> Still think that method should not be exposed as public interface method, though :)

Hmm.. `FilterCollection::loadFilter()` is used by `FilterCollection` itself only so I guess making it protected and removing from `FilterCollectionInterface` is fine.

**Comment by dereuromark** [comment]
Looks nice!

**Pull request merged**